server/plugins/LFUGens.cpp

```diff
@@ -3197,56 +3200,23 @@ void IEnvGen_Ctor(IEnvGen* unit) {
         SETCALC(IEnvGen_next_k);
     }
 
-    // pointer, offset
-    // initlevel, numstages, totaldur,
-    // [dur, shape, curve, level] * numvals
     int numStages = (int)IN0(3);
     int numvals = numStages * 4; // initlevel + (levels, dur, shape, curves) * stages
-    float offset = unit->m_offset = IN0(1);
-    float point = unit->m_pointin = IN0(0) - offset;
     unit->m_envvals = (float*)RTAlloc(unit->mWorld, (int)(numvals + 1.) * sizeof(float));
     ClearUnitIfMemFailed(unit->m_envvals);
 
+    // fill m_envvals with the values
     unit->m_envvals[0] = IN0(2);
-    //  Print("offset of and initial  values %3,3f, %3.3f\n", offset, unit->m_envvals[0]);
-    // fill m_envvals with the values;
     for (int i = 1; i <= numvals; i++) {
         unit->m_envvals[i] = IN0(4 + i);
-        //      Print("val for: %d, %3.3f\n", i, unit->m_envvals[i]);
     }
 
-    //  float out = OUT0(0);
-    float totalDur = IN0(4);
-    float level = 0.f;
-    float newtime = 0.f;
-    int stage = 0;
-    float seglen = 0.f;
-    if (point >= totalDur) {
-        unit->m_level = level = unit->m_envvals[numStages * 4]; // grab the last value
-    } else {
-        if (point <= 0.0) {
-            unit->m_level = level = unit->m_envvals[0];
-        } else {
-            float segpos = point;
-            // determine which segment the current time pointer needs calculated
-            for (int j = 0; point >= newtime; j++) {
-                seglen = unit->m_envvals[(j * 4) + 1];
-                newtime += seglen;
-                segpos -= seglen;
-                stage = j;
-            }
-
-            segpos = segpos + seglen;
-            float begLevel = unit->m_envvals[(stage * 4)];
-            int shape = (int)unit->m_envvals[(stage * 4) + 2];
-            int curve = (int)unit->m_envvals[(stage * 4) + 3];
-            float endLevel = unit->m_envvals[(stage * 4) + 4];
-            float pos = (segpos / seglen);
+    unit->m_offset = IN0(1);
+    unit->m_pointin = uninitializedControl; // trigger calculation in next_k
+    unit->m_level = uninitializedControl; // will be set in next_k
 
-            GET_ENV_VAL
-        }
-    }
-    OUT0(0) = level;
+    IEnvGen_next_k(unit, 1);
+    // state is unchanged by next_k, so no need to reset state after init
```

Comment:
great, thank you! however it's still a bit incorrect: next_k updates state, it's the output that doesn't need to be calculated again.... I'm trying to find a more precise wording... maybe something like
```cpp
// we can avoid resetting internal state: 
// since env is not updated in _next, output depends only on pointin. 
// if pointin doesn't change, m_level cached by _next is valid.
```

Reply:
how about:

The internal state doesn't need to be reset here: `m_pointin` determines `m_level` and is unchanged between initialization an the first 'next_' call.

Reply:
I just thought we could be more direct and concise.

> since env is not updated in _next

I'm not sure what this refers to. env isn't a variable, and the envelope itself of course isn't updated between init and next.

> if pointin doesn't change

we know it doesn't change, so we can just say that because it doesn't change, m_level doesn't either.

How about
```
// No need to reset internal state: 
// The output depends only on pointin, which doesn't change between
// initialization and the `next_` call. So the cached m_level is valid.
```

Thanks for working through these details.

Reply:
Thanks for bearing with me :D 
I think your last version is fine, and anyways what we want to signal is that we are skipping the reset on purpose. For more detailed understanding, one has to read the rest of the code anyways.

For the records, pointin could change between init and _next, in case it's from a UGen which `y(0) != presample` (but if I'm not wrong, we consider this behavior a bug)

Reply:
Ok thanks!

> in case it's from a UGen which y(0) != presample

Yes this is considered a bug, and in fact it's the goal of this whole initialization fix effort to resolve it :)

Reply:
Updated!